src/Layers/xrRender/r__dsgraph_render.cpp

```diff
@@ -46,7 +47,7 @@ void R_dsgraph_structure::r_dsgraph_render_graph(u32 _priority)
         auto& map = mapNormalPasses[_priority][iPass];
 
         map.get_any_p(nrmPasses);
-        std::sort(nrmPasses.begin(), nrmPasses.end(), cmp_pass<mapNormal_T::value_type*>);
+        std::stable_sort(nrmPasses.begin(), nrmPasses.end(), cmp_pass<mapNormal_T::value_type*>);
```

Comment:
Why sort algorithm and comparator was changed? What are the results of these changes?

Reply:
As far as I understand, the comparator contained logical error leading to UB in resulting sort algorithm, causing crashes (somehow) on my Mac machine.
The error is that it returns true for two equal arguments, that means cmp(a, b) == cmp(b, a) if a.equal(b) or, in the other words, a < b && b < a is true. Such comparator behavior doesn't meet the C++ standard criteria https://en.cppreference.com/w/cpp/named_req/Compare and doesn't make much sense (how can we order two items if one orders before another, but another is also before the one).
Looking at the code I made a guess that the intent was actually to sort SPass'es based on ssa, but doesn't touch equal ones (though I didn't figure out, why there can be equal SPass'es and can they have different ssa). Fell free to correct me if I got it wrong.
As for stable sort, I just followed the intuition that it's always better not to mess up rendering order from frame to frame (which may cause kind of flickering). So, knowing there can be equal items in the nrmPasses I decided to elide possible sorting instability factor.

Reply:
Maybe it's better to move this particular change to another pull request... Or, @vTurbine, can you tell if this change in comparator and stable sort is ok or even better than what we have currently?

Original engine code was sorting the scene graph in a special way to minimize state changes. I simplified the code to make SPass look like PSO (so we can add DX12/Vulkan in future), but this optimization was lost... but I tried to make comparator sort like in the original code.
Originally, sorting was performed with two criterias instead of just one: state change and SSA.
You can look at the original sorting code here: 5aeb5a64022628626688758d8a613d095511bddd. Maybe, you will be able to

Reply:
The comment seems to be truncated, what was the question "Maybe, you will be able to")?

Reply:
> Maybe it's better to move this particular change to another pull request

Sure, it's possible. But this change is absolutely required to run the game on macOS: without it the game crashes right at the beginning, rendering the first frame.

Reply:
>  but I tried to make comparator sort like in the original code

As far as I can see the comparator is not like in the original code, but rather nonsense because of 

>  it returns true for two equal arguments, that means cmp(a, b) == cmp(b, a) if a.equal(b) or, in the other words, a < b && b < a is true

I'm pretty much surprised how it even works on windows and what are the resulting order of the items.

Reply:
@Lnd-stoL, so, with @vTurbine we have figured that we don't need stable sort here and it's better to return generic sort. Comparator change is ok and enough! 🙂